src/Sylius/Behat/Context/Setup/PromotionContext.php

```diff
@@ -93,6 +103,26 @@ public function thereIsPromotion($promotionName)
     }
 
     /**
+     * @Given the store has promotion :promotionName with coupon :couponCode
+     */
+    public function thereIsPromotionWithCoupon($promotionName, $couponCode)
+    {
+        /** @var CouponInterface $coupon */
+        $coupon = $this->couponFactory->createNew();
+        $coupon->setCode($couponCode);
+
+        $promotion = $this->testPromotionFactory
+            ->createForChannel($promotionName, $this->sharedStorage->get('channel'))
+        ;
+        $promotion->addCoupon($coupon);
+        $promotion->setCouponBased(true);
+
+        $this->promotionRepository->add($promotion);
```

Comment:
You've added something to repository, but you didn't flush your changes. It can generate some strange behaviour, so I would suggest to add flush at the end of this function